### PR TITLE
aws: switch to aws-sdk

### LINKED ISSUE #12371: Switch to aws rust sdk
## Development Task

rusoto is not maintained and AWS has released its own SDK. https://github.com/awslabs/aws-sdk-rust.

Although it claims it's still not production ready, but according to https://github.com/rusoto/rusoto/issues/1651, community has started the transition. We may also give it a try.

### PR DESCRIPTION
**Pull request by akoshchiy**
### What is changed and how it works?

Issue Number: Close #12371 

What's Changed:

```commit-message
* switch kms to aws_sdk lib
* switch s3 to aws_sdk lib 
```

### Check List 

Tests 

- Unit test
- Manual test (tested with minio and localstack, see s3::tests::test_real_s3_storage and kms::tests::test_aws_kms_localstack)

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

```release-note
Please add a release note.

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

If you don't think this PR needs a release note then fill it with None.
If this PR will be picked to release branch, then a release note is probably required.
```

### CODE REVIEW

**Comment by akoshchiy** [comment]
aws-sdk maintainers have added virtual-host support for s3 recently, so I've added virtual host support too. Also, I've completely removed rusoto dependency from project.

File: components/cloud/aws/Cargo.toml
Comment (akoshchiy) [reply components/cloud/aws/Cargo.toml]: It's for mocking requests in tests. Will move to dev-dependencies.

**Comment by 3pointer** [comment]
After rusoto entered maintenance mode and the official SDK announced its availability for some time, after internal discussions, we hope to merge this PR as long as the conflicts are resolved. After that, we plan to submit another PR to keep the previous SDK for compatibility, and we will do more testing, and then pick this PR to other LTS versions.

@akoshchiy Could you please resolve the existing conflicts so that we can further improve upon this basis?

**Comment by 3pointer** [comment]
> After rusoto entered maintenance mode and the official SDK announced its availability for some time, after internal discussions, we hope to merge this PR as long as the conflicts are resolved. After that, we plan to submit another PR to keep the previous SDK for compatibility, and we will do more testing, and then pick this PR to other LTS versions.
> 
> @akoshchiy Could you please resolve the existing conflicts for us to merge this PR so that we can further improve upon this basis?

**Comment by akoshchiy** [comment]
> Could you please resolve the existing conflicts for us to merge this PR so that we can further improve upon this basis?

Sure :)

**Comment by akoshchiy** [comment]
Trying to update to latest aws crate versions.

**Comment by YuJuncen** [comment]
Thanks for your contributing! You may update your branch and then we can probably pass the `clippy-*` checks. Also you may sign-off your last commit.

I'm reviewing this PR on progress. Also sorry for the late reaction.

**Review by YuJuncen — APPROVED** [review]

**Comment by akoshchiy** [comment]
Seems that aws-sdk is using some of the banned dependencies.

**Comment by 3pointer** [comment]
> Seems that the aws-sdk is using some of the banned dependencies.

We are conducting internal discussions and will analyze each of these dependent libraries to see if adjustments can be made.

**Comment by 3pointer** [comment]
> > Seems that the aws-sdk is using some of the banned dependencies.
> 
> We are conducting internal discussions and will analyze each of these dependent libraries to see if adjustments can be made.

These banned dependency is used to achieve FIPS 140-2 compliance. we have submit a ticket to AWS to see whether they can provide a FIPS support sdk. 
BTW someone has met the same issue: https://github.com/awslabs/aws-sdk-rust/issues/966. 

Let's wait for the ticket feedback.

**Comment by kennytm** [comment]
cc #16443

**Comment by kennytm** [comment]
https://github.com/awslabs/aws-sdk-rust/issues/966#issuecomment-2075177698

> this *\*[FIPS support]* is now supported if you use our experimental support for Hyper 1.0: https://github.com/awslabs/aws-sdk-rust/blob/main/sdk/s3/tests/hyper-10.rs

See also https://github.com/awslabs/aws-sdk-rust/issues/1050#issuecomment-2130057612 for how to use the hyper 1.0 client. 

The feature is titled "experimental" though.

**Comment by YuJuncen** [comment]
Hi, I'm going to push some commits from https://github.com/YuJuncen/tikv/tree/aws-sdk-switch.

It solved the FIPS problem by replacing `rustls` with native TiKV openssl and resolved the conflicting.

**Comment by YuJuncen** [comment]
(Oops... some of commits are not signed-off... You may squash them or just let me know whether you are minding that I force-pushing to your branch...)

**Comment by akoshchiy** [comment]
> (Oops... some of commits are not signed-off... You may squash them or just let me know whether you are minding that I force-pushing to your branch...)

Thanks for the help! Yeah, feel free to force-pushing to the branch.

**Comment by YuJuncen** [comment]
I tried to sign-off and rebase commits... However it seems after many merge commits, it is pretty hard to rebase to a signed-off history. Let me check whether we have some workarounds when merging...

**Comment by YuJuncen** [comment]
cc @3pointer PTAL

**Comment by YuJuncen** [comment]
I have added a tiny fix over the previous commit that allows `hyper_tls` to load the local root certs.

**Comment by YuJuncen** [comment]
I have just updated the version of AWS SDK and adapted the new traits added at https://github.com/tikv/tikv/pull/17285

**Comment by YuJuncen** [comment]
@akoshchiy Hi, we run some tests internally and have fixed an issue that the client complains no http client find when fetching the AK/SK from IDMSv2 / Pod ID.

Also, it seems I cannot trivially rebase the history. You may need to squash the history (I cannot do that because if I did the author of the commit will become me.), by:

1. `git pull origin master --no-rebase`.
2. `git reset origin/master`.
3. `git commit -sam "..."`. 
4. `git push -f`.

**Comment by akoshchiy** [comment]
> @akoshchiy Hi, we run some tests internally and have fixed an issue that the client complains no http client find when fetching the AK/SK from IDMSv2 / Pod ID. I guess this PR should almost be ready to be merged. Also cc @3pointer and @Leavrth.
> 
> Also, it seems I cannot trivially rebase and sign off the history. You may need to squash the history (I cannot do that because if I did the author of this commit becomes me.), by:
> 
> 0. `git pull` (So my new changes will be included.)
> 1. `git pull origin master --no-rebase`
> 2. `git reset origin/master`
> 3. `git commit -sam "..."`
> 4. `git push -f`

Done! Thanks again for the help!

**Comment by YuJuncen** [comment]
/cc @overvenus 
Please take a look about the modifications over `deny.toml`, thanks!

File: Cargo.toml
Comment (overvenus) [inline Cargo.toml:208]:
```suggestion
snappy-sys = { git = "https://github.com/tikv/rust-snappy.git", branch = "static-link" }
```

Please keep tikv/rust-snappy.

**Review by overvenus — COMMENTED** [review]
Thank you! `deny.toml` LGTM.

**Review by 3pointer — COMMENTED** [review]
rest LGTM

**Review by 3pointer — APPROVED** [review]

File: components/cloud/aws/src/s3.rs
Comment (BornChanger) [reply components/cloud/aws/src/s3.rs:228]: Who is responsible to clarify that?

Comment (3pointer) [reply components/cloud/aws/src/s3.rs:228]: From aws document, https://docs.aws.amazon.com/STS/latest/APIReference/welcome.html
> AWS recommends using Regional AWS STS endpoints instead of the global endpoint to reduce latency, build in redundancy, and increase session token validity. 

It's better to use regional STS here. But

> Most AWS Regions enable operations in all AWS services by default. These Regions automatically activate for use with AWS STS. Some Regions, such as Asia Pacific (Hong Kong), must be manually enabled.

So, we should let user make sure the S3 bucket region has enabled STS service. Consider this make sense. It's okay to add this assumption. And we'd better test the behavior after this PR merged.

**Review by Leavrth — APPROVED** [review]

**Comment by BornChanger** [comment]
@akoshchiy please fix the failed ut `test_s3_storage_multi_part`

**Comment by wuhuizuo** [comment]
/release-note-edit ```release-note Switch to the official AWS SDK for S3 operations, replacing Rusoto for enhanced support ```

**Comment by purelind** [comment]
/release-note-edit ```release-note Switch to the official AWS SDK for S3 operations, replacing Rusoto for enhanced support ```

**Comment by wuhuizuo** [comment]
/assign @jebter

Help to modify the release note.

**Pull request merged**
